src/exercises/2-4-State_encapsulation/Readme.md

```diff
@@ -0,0 +1,71 @@
+In this exercise we’ll create a custom React Hook to manage the state of our Dialog component.
+
+## 2.4 Encapsulating state in a React Hook
+
+In the previous exercise we’ve created a collection of stateless components. This gave us almost unlimited flexibility on how to compose these components. But since this components are stateless, we still have to manage the state (in our case the state of the modal opening) ourselves, and we have to repeate this code everty time we want to create a modal.
+
+To solve this issue, we’re going to create a custom React Hook that encapsulates the state and returns its current value and a function to change this value. We’ll also need to move the condition from the caller code to our container `Dialog` component.
+
+For example, we have a disclosure component, that consits of some content and a button that shows or hides the content. We can implement the state on the call site:
+
+```jsx static
+const [isExpanded, setIsExpanded] = React.useState(false);
+<>
+  <DisclosureButton onClick={() => setIsExpanded(x => !x)}>
+    Toggle
+  </DisclosureButton>
+  {isExpanded && <DisclosureContent>Content</DisclosureContent>}
+</>;
+```
+
+Or we can encapsulate the state management in a custom Hook, and spread the returned value to each primitive:
+
+```jsx static
+function useDisclosureState({
+  expanded: defaultExpanded = false
+} = {}) {
+  const [expanded, setIsExpanded] = React.useState(defaultExpanded);
+  return {
+    expanded,
+    onToggle: () => setIsExpanded(x => !x)
+  };
+}
+
+const disclosure = useDisclosureState({ expanded: false });
+<>
+  <DisclosureButton {...disclosure}>Toggle</DisclosureButton>
+  <DisclosureContent {...disclosure}>Content</DisclosureContent>
+</>;
+```
+
+The benefit is that we don’t have to remember the API of each primitive component, and can spread the same props to each component, and let them deal with it. However, we still have access to all the values and function, and could use them in our custom components. We also reduce code duplication and make the behaviour more consistent, which is especially important for components with more complex behavior.
+
+**Note:** This exercise is inspired by the [Reakit Dialog component](https://reakit.io/docs/dialog/).
+
+**Warning:** The implementation here is very basic, and only an example of the technique. It shouldn’t be used in production as is to create modal dialogs because we didn’t consider accessibility. Check our the [Reakit Dialog component](https://reakit.io/docs/dialog/#accessibility) to have an idea of what’s necessary to make a dialog accessible.
+
+### The result
+
+The result should look like this:
+
+```jsx {"file": "final/Dialog.md", "noeditor": true}
+```
+
+### The task
+
+1. Create a custom React Hook to manage Dialog opening state. The Hook should return the current opening state and function to show and hide the dialog.
```

Comment:
Hmmm, didn't you just tell them what to do? Maybe we can explain the technique on another example and let them build it here and after they are done we could repeat and refactor our with live coding?

Reply:
Do you have better ideas? I've tried with Tabs but it's way too complex, and obscures the idea ;-/

Reply:
Not really. Please come up with one?

Reply:
Let's do that later

Reply:
I'll try to think about something that isn't show/hide of some kind and not too complex like tabs 🦛